kafka-connect-fitbit-source/src/main/java/org/radarbase/connect/rest/fitbit/converter/FitbitAvroConverter.java

```diff
@@ -98,6 +95,22 @@ public FitbitAvroConverter(AvroData avroData) {
         .collect(Collectors.toList());
   }
 
+  private boolean validateRecord(FitbitRestRequest request, TopicData record) {
+    if (record == null) {
+      return false;
+    }
+    Instant endDate = request.getUser().getEndDate();
+    if (endDate == null) {
+      return true;
+    }
+    Field timeField = record.value.getSchema().getField("time");
+    if (timeField != null) {
+      long time = (long) (((Double)record.value.get(timeField.pos()) * 1000.0));
+      return Instant.ofEpochMilli(time).isBefore(endDate);
+    }
+    return true;
+  }
```

Comment:
This will filter the records after the request is made and data is received? But why are we even making requests after the end date?
The problem right now is that the connector tries to request after the end date. For that, it needs to refresh the token, but user's credentials are not correct anymore (they changed the password after exiting the study) and this results in a lot of failed refresh token requests, even though the data is complete and they should not be queried anymore.

Reply:
This avoids data being pulled with an open-ended api call (only based on lastUpdate), data could be added that does not match the date criteria.

To fix your particular problem, I would advise adding a `User.isAuthorized` method to `User` and add to the `isComplete` method `&& isAuthorized()`.

Alternatively, in `FitbitPollingRoute.nextPoll`, you could change the condition
```java
     if (offset.isAfter(user.getEndDate())) {
      return nearFuture();
    }
```

to

```java
     if (offset.isAfter(user.getEndDate().minus(getEndDateThreshold()))) {
      return nearFuture();
    }


    protected Duration getEndDateThreshold() {
       return Duration.ofHours(1);
    }
```
Still if there's no data in that last hour, the connector will keep trying to fetch that final hour.

Reply:
> To fix your particular problem, I would advise adding a User.isAuthorized method to User and add to the isComplete method && isAuthorized().

How do we check if the user is not authorised?, since even when refresh token request fails, the authorizer still shows the subject as Authorized=true.

Reply:
> How do we check if the user is not authorised?, since even when refresh token request fails, the authorizer still shows the subject as Authorized=true.

That would be good to evaluate. It should not.

Reply:
Oh. We should check that indeed. It could be that the condition is not set to false when the request is not successful.

Reply:
I think we have an add explicit check for response status and update the user's isAuthorised status. 
https://github.com/RADAR-base/RADAR-Rest-Source-Auth/blob/dev/authorizer-app-backend/src/main/java/org/radarbase/authorizer/resources/RestSourceUserResource.kt#L165
At the moment, we don't. 
https://github.com/RADAR-base/RADAR-Rest-Source-Auth/blob/dev/authorizer-app-backend/src/main/java/org/radarbase/authorizer/doa/RestSourceUserRepositoryImpl.kt#L53